### PR TITLE
chore(sdk): Add annotations for new-styled KFP component

### PR DESCRIPTION
**Pull request by numerology**
**Description of your changes:**

Add a component annotation for new-styled KFP component when users specify such. So that downstream logic can be conditioned on this.

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: sdk/python/kfp/containers/_component_builder.py [hunk 1]
```diff
@@ -13,21 +13,22 @@
...
+from typing import Callable, List, Optional
 
 from deprecated.sphinx import deprecated
 
 from ..components._components import _create_task_factory_from_component_spec
 from ..components._python_op import _func_to_component_spec
 from ._container_builder import ContainerBuilder
+from kfp.components import _structures
+
+NEW_STYLED_COMPONENT_ANNOTATION = 'pipelines.kubeflow.org/new_styled_component'
```
Comment (neuromage) [inline sdk/python/kfp/containers/_component_builder.py:30]: How about:

`V2_COMPONENT_ANNOTATION = "pipelines.kubeflow.org/component_v2` 

?

Comment (numerology) [reply sdk/python/kfp/containers/_component_builder.py]: Done

**Review by numerology — COMMENTED** [review]
Thanks

**Comment by chensun** [comment]
/lgtm

Thanks!

File: sdk/python/kfp/containers/_component_builder.py [hunk 2]
```diff
@@ -214,7 +235,24 @@ def build_python_component(component_func, target_image, base_image=None, depend
                                   ' and push the image to ' +
                                   target_image)
 
-  component_spec = _func_to_component_spec(component_func, base_image=base_image)
+  component_spec = _func_to_component_spec(
+      component_func, base_image=base_image)
+
+  if is_v2:
+    # Annotate the component to be a V2 one.
+    if getattr(component_spec, 'metadata', None) and getattr(
```
Comment (Ark-kun) [inline sdk/python/kfp/containers/_component_builder.py:243]: JFYI: `if getattr(component_spec, 'metadata', None)` can probably be replaced with just `if component_spec.metadata`. The structure field always exist (although they can be None).

Comment (Ark-kun) [reply sdk/python/kfp/containers/_component_builder.py:243]: Maybe we can try to make this code easier to understand:
```python
if not component_spec.metadata:
    component_spec.metadata = _structures.MetadataSpec()
if not component_spec.metadata.annotations:
    component_spec.metadata.annotations = {}

component_spec.metadata.annotations[V2_COMPONENT_ANNOTATION] = 'true'
```

Comment (numerology) [reply sdk/python/kfp/containers/_component_builder.py]: Good point! Done.

**Pull request merged**